src/operations.py

```diff
@@ -82,7 +85,11 @@ def run(self):
         log.debug('node %s running operation thread with operations %s',
             self.node_id, self.ops_to_run)
 
-        for op in self.ops_to_run:
+        # Add Heartbeat Action first
+        ops_to_run = copy.deepcopy(self.ops_to_run)
+        ops_to_run.insert(0, Op.Heartbeat)
```

Comment:
why not just append?

Reply:
I thought it will be better to check liveness (and handle properly such as update graph) before other actions

Reply:
This sort of overloads the usage of ops_to_run i.e. it's supposed to work in both test time and actual execution. We can work with this actually, seems better than having to maintain multiple ops queues.
Comment on the diff: Note that this approach would mean that we're NOT guaranteeing fairness. Operations that haven't been executed due to slow running of our Process can potentially be starved. Let's go with appending it to the end.

Reply:
sounds like appending is the better option

Reply:
Got it, refactoring to appending.